In [72]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

# PART 1: Load Data

In [73]:
#Importing the data
cell_phones = pd.read_csv("https://raw.githubusercontent.com/zalihat/cell_phone_dataset/master/cell_phones.csv")
population = pd.read_csv("https://raw.githubusercontent.com/zalihat/cell_phone_dataset/master/population1.csv")
country_code = pd.read_csv("https://raw.githubusercontent.com/open-numbers/ddf--gapminder--systema_globalis/b557bdc0693112f45c2d0330bb42a54b37713ac9/ddf--entities--geo--country.csv")

In [98]:
#Saving the datasets locally on my computer
#cell_phones.to_csv('cellphones.csv', index = False)
#population.to_csv('population.csv', index = False)
#country_code.to_csv('geo_country_code', index = False)

In [157]:
cell_phones.head()

Country  year  cell_phones_total  phone_per_person
0  Afghanistan  1960                0.0          1.230769
1  Afghanistan  1961                0.0          1.230769
2  Afghanistan  1962                0.0          1.230769
3  Afghanistan  1963                0.0          1.230769
4  Afghanistan  1964                0.0          1.230769

In [158]:
population.head()

Country  year  population
0  Afghanistan  1800     3280000
1  Afghanistan  1801     3280000
2  Afghanistan  1802     3280000
3  Afghanistan  1803     3280000
4  Afghanistan  1804     3280000

In [159]:
country_code.head()

geo g77_and_oecd_countries income_3groups        income_groups  \
0  abkh                 others            NaN                  NaN   
1   abw                 others    high_income          high_income   
2   afg                    g77     low_income           low_income   
3   ago                    g77  middle_income  lower_middle_income   
4   aia                 others            NaN                  NaN   

   is--country iso3166_1_alpha2 iso3166_1_alpha3  iso3166_1_numeric iso3166_2  \
0         True              NaN              NaN                NaN       NaN   
1         True               AW              ABW              533.0       NaN   
2         True               AF              AFG                4.0       NaN   
3         True               AO              AGO               24.0       NaN   
4         True               AI              AIA              660.0       NaN   

   landlocked  ...  longitude  main_religion_2008      Country  \
0         NaN  ...        NaN                 NaN     Abkhazia   
1   coastline  ...  -69.96667           christian        Aruba   
2  landlocked  ...   66.00000              muslim  Afghanistan   
3   coastline  ...   18.50000           christian       Angola   
4   coastline  ...  -63.05000           christian     Anguilla   

               un_sdg_ldc                       un_sdg_region un_state  \
0                     NaN                                 NaN    False   
1  un_not_least_developed  un_latin_america_and_the_caribbean    False   
2      un_least_developed        un_central_and_southern_asia     True   
3      un_least_developed               un_sub_saharan_africa     True   
4  un_not_least_developed  un_latin_america_and_the_caribbean    False   

   unicef_region unicode_region_subtag world_4region        world_6region  
0            NaN                   NaN        europe  europe_central_asia  
1            NaN                    AW      americas              america  
2             sa                    AF          asia           south_asia  
3            ssa                    AO        africa   sub_saharan_africa  
4            NaN                    AI      americas              america  

[5 rows x 21 columns]

# PART 2: Join Data
* Join the ‘Cell phones’ and ‘population’ dataframe (With an inner join on ‘Country’ and ‘year’)

In [160]:
phone_population = pd.merge(cell_phones, population, how = 'inner') 

In [161]:
phone_population.head()

Country  year  cell_phones_total  phone_per_person  population
0  Afghanistan  1960                0.0          1.230769     9000000
1  Afghanistan  1961                0.0          1.230769     9170000
2  Afghanistan  1962                0.0          1.230769     9350000
3  Afghanistan  1963                0.0          1.230769     9540000
4  Afghanistan  1964                0.0          1.230769     9740000

* In the ‘Geo country codes’ rename the columns{‘country to ‘geo’, and ‘name’ to ‘Country’

In [162]:
country_code = country_code.rename(columns={'country':'geo', 'name':'Country'})

In [163]:
country_code.head()

geo g77_and_oecd_countries income_3groups        income_groups  \
0  abkh                 others            NaN                  NaN   
1   abw                 others    high_income          high_income   
2   afg                    g77     low_income           low_income   
3   ago                    g77  middle_income  lower_middle_income   
4   aia                 others            NaN                  NaN   

   is--country iso3166_1_alpha2 iso3166_1_alpha3  iso3166_1_numeric iso3166_2  \
0         True              NaN              NaN                NaN       NaN   
1         True               AW              ABW              533.0       NaN   
2         True               AF              AFG                4.0       NaN   
3         True               AO              AGO               24.0       NaN   
4         True               AI              AIA              660.0       NaN   

   landlocked  ...  longitude  main_religion_2008      Country  \
0         NaN  ...        NaN                 NaN     Abkhazia   
1   coastline  ...  -69.96667           christian        Aruba   
2  landlocked  ...   66.00000              muslim  Afghanistan   
3   coastline  ...   18.50000           christian       Angola   
4   coastline  ...  -63.05000           christian     Anguilla   

               un_sdg_ldc                       un_sdg_region un_state  \
0                     NaN                                 NaN    False   
1  un_not_least_developed  un_latin_america_and_the_caribbean    False   
2      un_least_developed        un_central_and_southern_asia     True   
3      un_least_developed               un_sub_saharan_africa     True   
4  un_not_least_developed  un_latin_america_and_the_caribbean    False   

   unicef_region unicode_region_subtag world_4region        world_6region  
0            NaN                   NaN        europe  europe_central_asia  
1            NaN                    AW      americas              america  
2             sa                    AF          asia           south_asia  
3            ssa                    AO        africa   sub_saharan_africa  
4            NaN                    AI      americas              america  

[5 rows x 21 columns]

* Then, select the ‘geo’ and ‘Country’ columns from the ‘geo_country’ dataframe, and join with your ‘merged population and cell phone’ data.

In [164]:
geo_phn_pop = pd.merge(phone_population, country_code, how = 'inner')

In [165]:
geo_phn_pop.head()

Country  year  cell_phones_total  phone_per_person  population  geo  \
0  Afghanistan  1960                0.0          1.230769     9000000  afg   
1  Afghanistan  1961                0.0          1.230769     9170000  afg   
2  Afghanistan  1962                0.0          1.230769     9350000  afg   
3  Afghanistan  1963                0.0          1.230769     9540000  afg   
4  Afghanistan  1964                0.0          1.230769     9740000  afg   

  g77_and_oecd_countries income_3groups income_groups  is--country  ...  \
0                    g77     low_income    low_income         True  ...   
1                    g77     low_income    low_income         True  ...   
2                    g77     low_income    low_income         True  ...   
3                    g77     low_income    low_income         True  ...   
4                    g77     low_income    low_income         True  ...   

  latitude longitude  main_religion_2008          un_sdg_ldc  \
0     33.0      66.0              muslim  un_least_developed   
1     33.0      66.0              muslim  un_least_developed   
2     33.0      66.0              muslim  un_least_developed   
3     33.0      66.0              muslim  un_least_developed   
4     33.0      66.0              muslim  un_least_developed   

                  un_sdg_region  un_state  unicef_region  \
0  un_central_and_southern_asia      True             sa   
1  un_central_and_southern_asia      True             sa   
2  un_central_and_southern_asia      True             sa   
3  un_central_and_southern_asia      True             sa   
4  un_central_and_southern_asia      True             sa   

  unicode_region_subtag world_4region world_6region  
0                    AF          asia    south_asia  
1                    AF          asia    south_asia  
2                    AF          asia    south_asia  
3                    AF          asia    south_asia  
4                    AF          asia    south_asia  

[5 rows x 25 columns]

In [166]:
col = phone_population.columns

cols = [] 
for i in col:
    cols.append(i)
    
cols.append('geo')
cols

['Country',
 'year',
 'cell_phones_total',
 'phone_per_person',
 'population',
 'geo']

In [167]:
#Selecting only the columns we need
data = geo_phn_pop[cols]

In [168]:
data.head()

Country  year  cell_phones_total  phone_per_person  population  geo
0  Afghanistan  1960                0.0          1.230769     9000000  afg
1  Afghanistan  1961                0.0          1.230769     9170000  afg
2  Afghanistan  1962                0.0          1.230769     9350000  afg
3  Afghanistan  1963                0.0          1.230769     9540000  afg
4  Afghanistan  1964                0.0          1.230769     9740000  afg

# PART THREE: Make Features
* Calculate the number of cell phones per person, and add this column onto your dataframe.

In [172]:
data['phone_per_person']  = data['cell_phones_total'] / data['population']

<ipython-input-172-acce64e7721f>:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['phone_per_person']  = data['cell_phones_total'] / data['population']


In [173]:
data.head()

Country  year  cell_phones_total  phone_per_person  population  geo
0  Afghanistan  1960                0.0               0.0     9000000  afg
1  Afghanistan  1961                0.0               0.0     9170000  afg
2  Afghanistan  1962                0.0               0.0     9350000  afg
3  Afghanistan  1963                0.0               0.0     9540000  afg
4  Afghanistan  1964                0.0               0.0     9740000  afg

* Assert: If you’ve calculated correctly you’ll get ‘1.230769’ cell phones per person in the United States in 2017.
* Checking to see if I got this right

In [178]:
data[(data['Country'] == 'United States') & (data['year'] == 2017)]

Country  year  cell_phones_total  phone_per_person  population  \
11157  United States  2017        400000000.0          1.230769   325000000   

       geo  
11157  usa

* As we can see above, our calculation is right, the number of phones per person in the United States in 2017 = 1.230769

# PART FOUR: Process Data
* Use the ‘.describe()’ method to Generate descriptive statistics and describe your dataframe's numeric columns, and then its non-numeric columns.

In [180]:
data.describe(include='all')

Country          year  cell_phones_total  phone_per_person  \
count       11640  11640.000000       1.164000e+04      11640.000000   
unique        194           NaN                NaN               NaN   
top     Guatemala           NaN                NaN               NaN   
freq           60           NaN                NaN               NaN   
mean          NaN   1989.500000       7.875722e+06          0.235536   
std           NaN     17.318846       5.572496e+07          0.433303   
min           NaN   1960.000000       0.000000e+00          0.000000   
25%           NaN   1974.750000       0.000000e+00          0.000000   
50%           NaN   1989.500000       0.000000e+00          0.000000   
75%           NaN   2004.250000       5.615000e+05          0.233667   
max           NaN   2019.000000       1.730000e+09          2.126068   

          population    geo  
count   1.164000e+04  11640  
unique           NaN    194  
top              NaN    bgr  
freq             NaN     60  
mean    2.710307e+07    NaN  
std     1.086187e+08    NaN  
min     4.380000e+03    NaN  
25%     1.117500e+06    NaN  
50%     5.100000e+06    NaN  
75%     1.590000e+07    NaN  
max     1.430000e+09    NaN

# Numeric Variables/Columns.
* There are 4 Numerical variables in our dataset: 'year', 'cell_phones_total', 'phone_per_person', and 'population'

### year
* The variable 'year' is a numerical variable, which contains years.
* The variable has 11640 rows, it has no missing values
* The oldest year is 1960, and the most recent year is 2019
* It has a mean of 1989.500000, and a median of 1989.500000, which tells us our this variable is symmetric. i.e. the data in this variable is fairly balanced.


### cell_phones_total
* The variable 'cell_phones_total' is a numerical variable, which contains data on the total number of cellphones in a particular country in a particular year
* The variable has 11640 rows, it has no missing values
* The minimum value here is 0.000, and the maximum value is 1.730000e+09
* Most of it's data is between the 75th percentile and the 100th percentile. 
* It has a mean of 7.875722e+06	, and a median of 0.0000, which tells us that it probably has a positive skew, We can confirm this by doing more analysis

### phone_per_person
* The variable 'phone_per_person' is a numerical variable that contains data about the average number of phones per person in a particular country in a particular year.
* The variable has 11640 rows, it has no missing values
* The minimum value here is 0.000, and on the average, the highest number of phones per person is 2.126068
* Most of it's data is between the 75th percentile and the 100th percentile. 
* It has a mean of 0.235536, and a median of 0.0000, which tells us that it probably has a positive skew, We can confirm this by doing more analysis

### population
* The variable 'population' is a numerical variable that contains data about the total number of people in a country in a particular year.
* The variable has 11640 rows, it has no missing values
* The country with the smallest population has a population of 4.380000e+03	, and the country with the highest has a population of 1.430000e+09.
* Most of it's data is between the 75th percentile and the 100th percentile. 
* It has a mean 2.710307e+07, and a median of 5.100000e+06, which tells us that it probably has a positive skew, We can confirm this by doing more analysis

# Categorical Variables/Columns.
* There are two categorical variables in our dataset: 'Country' and 'geo'

### Country
* 'Country' is a categorical variable.
* From the method above, we can see that the variable, **Country** has 11640 rows, it has no missing values.
* It has 194 unique values i.e. 194 countries
* The country with the most occurence is Guatemala, which occurs 60 times.

### geo
* 'geo' is a categorical variable.
* From the statistics above, we can see that the variable, **geo** has 11640 rows, it has no missing values.
* It has 194 unique values i.e. 194 geo codes
* The geo code  with the most occurence is Guatemala, which occurs 60 times.

* In 2017, what were the top 5 countries with the most cell phones total?
* We will get this by aggregating through our data.

In [196]:
#filtering our data to show only data in the year 2017.
data_2017 = data[data['year'] == 2017]
data_2017.head()

Country  year  cell_phones_total  phone_per_person  population  geo
57   Afghanistan  2017         23900000.0          0.658402    36300000  afg
117      Albania  2017          3630000.0          1.260417     2880000  alb
177      Algeria  2017         45800000.0          1.106280    41400000  dza
237      Andorra  2017            80300.0          1.042857       77000  and
297       Angola  2017         13300000.0          0.446309    29800000  ago

In [214]:
#Grouping each country together, and getting the total cell phones per country in 2017
data_2017_grouped = data_2017[['Country', 'cell_phones_total']].groupby('Country').sum()
data_2017_grouped.head()

cell_phones_total
Country                       
Afghanistan         23900000.0
Albania              3630000.0
Algeria             45800000.0
Andorra                80300.0
Angola              13300000.0

In [215]:
#Now sorting this data from highest to lowest, and getting the top 5 countries with the highest number of cell phones in 2017
data_2017_grouped_sorted = data_2017_grouped.sort_values(by='cell_phones_total', ascending = False)

data_2017_grouped_sorted.head(5)

cell_phones_total
Country                         
China               1.470000e+09
India               1.170000e+09
Indonesia           4.350000e+08
United States       4.000000e+08
Russia              2.270000e+08

* The countries with the highest number of cell phones in 2017 are China, India, Indonesia, United States, and Russia, in that order.